File: lib/src/base/mongol_text_metrics.dart
// Copyright 2014 The Flutter Authors.
// Copyright 2020 Suragch.
// All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:math' show max, min;
import 'package:flutter/foundation.dart' show immutable;
import 'package:flutter/services.dart' show TextAffinity, TextPosition;
import 'package:flutter/widgets.dart' show InlineSpan, Offset, Rect;
import 'mongol_paragraph.dart';
import 'mongol_text_tools.dart';

/// 光标度量信息的基类
/// Base class for caret metrics information.
/// Base class for caret metrics.
@immutable
sealed class CaretMetrics {}

/// 光标位于非空行的中的度量信息
/// Caret metrics for non-empty lines.
final class LineCaretMetrics implements CaretMetrics {
  const LineCaretMetrics({required this.offset, required this.fullWidth});

  final Offset offset;

  final double fullWidth;
}

/// 光标在空行处的度量
/// Caret metrics for empty lines.
final class EmptyLineCaretMetrics implements CaretMetrics {
  const EmptyLineCaretMetrics({required this.lineHorizontalOffset});

  final double lineHorizontalOffset;
}

/// 光标度量计算器
/// Calculates caret metrics for text positions.
class CaretMetricsCalculator {
  static const int _zeroWidthJoinerCodeUnit = 0x200d;
  static const int _newlineCodeUnit = 0x0A;
  static const EmptyLineCaretMetrics _emptyCaretMetrics =
      EmptyLineCaretMetrics(lineHorizontalOffset: 0);

  CaretMetrics _cachedCaretMetrics = _emptyCaretMetrics;

  TextPosition? _lastQueriedCaretPosition;
  String? _lastQueriedPlainText;
  MongolParagraph? _lastQueriedParagraph;
  InlineSpan? _lastQueriedText;

  bool _canUseCached(
    TextPosition position,
    String plainText,
    MongolParagraph paragraph,
    InlineSpan text,
  ) {
    return position == _lastQueriedCaretPosition &&
        identical(plainText, _lastQueriedPlainText) &&
        identical(paragraph, _lastQueriedParagraph) &&
        identical(text, _lastQueriedText);
  }

  CaretMetrics _cacheAndReturn(
    TextPosition position,
    String plainText,
    MongolParagraph paragraph,
    InlineSpan text,
    CaretMetrics metrics,
  ) {
    _lastQueriedCaretPosition = position;
    _lastQueriedPlainText = plainText;
    _lastQueriedParagraph = paragraph;
    _lastQueriedText = text;
    return _cachedCaretMetrics = metrics;
  }

  CaretMetrics compute(
    TextPosition position,
    String plainText,
    MongolParagraph paragraph,
    InlineSpan text,
  ) {
    if (_canUseCached(position, plainText, paragraph, text)) {
      return _cachedCaretMetrics;
    }

    final int offset = position.offset;
    final int plainTextLength = plainText.length;
    if (offset < 0 || offset > plainTextLength || plainTextLength == 0) {
      return _cacheAndReturn(
          position, plainText, paragraph, text, _emptyCaretMetrics);
    }

    final CaretMetrics? metrics = switch (position.affinity) {
      TextAffinity.upstream =>
        _getMetricsFromUpstream(offset, plainText, paragraph, text) ??
            _getMetricsFromDownstream(offset, plainText, paragraph, text),
      TextAffinity.downstream =>
        _getMetricsFromDownstream(offset, plainText, paragraph, text) ??
            _getMetricsFromUpstream(offset, plainText, paragraph, text),
    };

    return _cacheAndReturn(
      position,
      plainText,
      paragraph,
      text,
      metrics ?? _emptyCaretMetrics,
    );
  }

  /// 根据前一个字符获取光标度量
  /// Gets caret metrics from the previous character.
  CaretMetrics? _getMetricsFromUpstream(
    int offset,
    String plainText,
    MongolParagraph paragraph,
    InlineSpan text,
  ) {
    final int plainTextLength = plainText.length;
    if (plainTextLength == 0 || offset < 0 || offset > plainTextLength) {
      return null;
    }

    final int safeOffset = offset.clamp(0, plainTextLength);
    final int prevCodeUnit = plainText.codeUnitAt(max(0, safeOffset - 1));
    final int? nextCodeUnit = text.codeUnitAt(safeOffset);

    final bool needsSearch =
        _needsGraphemeExtendedSearch(prevCodeUnit, nextCodeUnit);

    final List<Rect> boxes = _findBoxesUpstream(
      paragraph,
      safeOffset,
      plainTextLength,
      needsSearch,
      stopAtNewline: prevCodeUnit == _newlineCodeUnit,
    );
    if (boxes.isEmpty) return null;

    final box = boxes.last;
    return prevCodeUnit == _newlineCodeUnit
        ? EmptyLineCaretMetrics(lineHorizontalOffset: box.right)
        : LineCaretMetrics(
            offset: Offset(box.left, box.bottom),
            fullWidth: box.right - box.left,
          );
  }

  CaretMetrics? _getMetricsFromDownstream(
    int offset,
    String plainText,
    MongolParagraph paragraph,
    InlineSpan text,
  ) {
    final int plainTextLength = plainText.length;
    if (plainTextLength == 0 || offset < 0 || offset > plainTextLength) {
      return null;
    }

    final int safeOffset = offset.clamp(0, plainTextLength);
    final int nextCodeUnit =
        plainText.codeUnitAt(min(safeOffset, plainTextLength - 1));

    final bool needsSearch = _needsGraphemeExtendedSearch(nextCodeUnit, null);
    final List<Rect> boxes = _findBoxesDownstream(
        paragraph, safeOffset, plainTextLength, needsSearch);

    if (boxes.isEmpty) return null;
    final box = boxes.first;
    return LineCaretMetrics(
      offset: Offset(box.left, box.top),
      fullWidth: box.right - box.left,
    );
  }

  List<Rect> _findBoxesUpstream(
    MongolParagraph paragraph,
    int offset,
    int plainTextLength,
    bool needsSearch, {
    required bool stopAtNewline,
  }) {
    int graphemeLength = needsSearch ? 2 : 1;

    while (true) {
      final int rangeStart = max(0, offset - graphemeLength);
      final List<Rect> boxes = paragraph.getBoxesForRange(rangeStart, offset);
      if (boxes.isNotEmpty) {
        return boxes;
      }

      if (!needsSearch || stopAtNewline || rangeStart == 0) {
        return const <Rect>[];
      }

      final int nextLength = min(plainTextLength, graphemeLength * 2);
      if (nextLength == graphemeLength) {
        return const <Rect>[];
      }
      graphemeLength = nextLength;
    }
  }

  List<Rect> _findBoxesDownstream(
    MongolParagraph paragraph,
    int offset,
    int plainTextLength,
    bool needsSearch,
  ) {
    int graphemeLength = needsSearch ? 2 : 1;
    final int maxSearchEnd = plainTextLength << 1;

    while (true) {
      final int rangeEnd = offset + graphemeLength;
      final List<Rect> boxes = paragraph.getBoxesForRange(offset, rangeEnd);
      if (boxes.isNotEmpty) {
        return boxes;
      }

      if (!needsSearch || rangeEnd >= maxSearchEnd) {
        return const <Rect>[];
      }

      final int nextLength = graphemeLength * 2;
      if (nextLength == graphemeLength) {
        return const <Rect>[];
      }
      graphemeLength = nextLength;
    }
  }

  bool _needsGraphemeExtendedSearch(int codeUnit, int? codeUnitAfter) {
    return MongolTextTools.isHighSurrogate(codeUnit) ||
        MongolTextTools.isLowSurrogate(codeUnit) ||
        codeUnitAfter == _zeroWidthJoinerCodeUnit ||
        MongolTextTools.isUnicodeDirectionality(codeUnit);
  }
}

File: jetpack_compose/src/main/kotlin/mongol/compose/core/MongolTextMetrics.kt
package mongol.compose.core

import kotlin.math.max
import kotlin.math.min

/**
 * Base type for caret metrics.
 */
sealed interface CaretMetrics

/**
 * Caret metrics for non-empty lines.
 */
data class LineCaretMetrics(
    val offset: Offset,
    val fullWidth: Float,
) : CaretMetrics

/**
 * Caret metrics for empty lines.
 */
data class EmptyLineCaretMetrics(
    val lineHorizontalOffset: Float,
) : CaretMetrics

/**
 * Calculates caret metrics for a text position.
 *
 * This is a direct Kotlin adaptation of Flutter's CaretMetricsCalculator from:
 * lib/src/base/mongol_text_metrics.dart
 */
class CaretMetricsCalculator {
    private companion object {
        const val ZERO_WIDTH_JOINER_CODE_UNIT = 0x200d
        const val NEWLINE_CODE_UNIT = 0x0A
        val EMPTY_CARET_METRICS = EmptyLineCaretMetrics(lineHorizontalOffset = 0f)
    }

    private var cachedCaretMetrics: CaretMetrics = EMPTY_CARET_METRICS

    private var lastQueriedCaretPosition: TextPosition? = null
    private var lastQueriedPlainText: String? = null
    private var lastQueriedParagraph: ParagraphBoxProvider? = null

    private fun canUseCached(
        position: TextPosition,
        plainText: String,
        paragraph: ParagraphBoxProvider,
    ): Boolean {
        return position == lastQueriedCaretPosition && plainText === lastQueriedPlainText && paragraph === lastQueriedParagraph
    }

    private fun cacheAndReturn(
        position: TextPosition,
        plainText: String,
        paragraph: ParagraphBoxProvider,
        metrics: CaretMetrics,
    ): CaretMetrics {
        lastQueriedCaretPosition = position
        lastQueriedPlainText = plainText
        lastQueriedParagraph = paragraph
        cachedCaretMetrics = metrics
        return metrics
    }

    fun compute(
        position: TextPosition,
        plainText: String,
        paragraph: ParagraphBoxProvider,
    ): CaretMetrics {
        if (canUseCached(position, plainText, paragraph)) {
            return cachedCaretMetrics
        }

        val offset = position.offset
        val plainTextLength = plainText.length
        if (offset < 0 || offset > plainTextLength || plainTextLength == 0) {
            return cacheAndReturn(position, plainText, paragraph, EMPTY_CARET_METRICS)
        }

        val metrics = when (position.affinity) {
            TextAffinity.UPSTREAM -> {
                getMetricsFromUpstream(offset, plainText, paragraph) ?: getMetricsFromDownstream(
                    offset,
                    plainText,
                    paragraph
                )
            }

            TextAffinity.DOWNSTREAM -> {
                getMetricsFromDownstream(offset, plainText, paragraph) ?: getMetricsFromUpstream(
                    offset,
                    plainText,
                    paragraph
                )
            }
        }

        return cacheAndReturn(position, plainText, paragraph, metrics ?: EMPTY_CARET_METRICS)
    }

    private fun getMetricsFromUpstream(
        offset: Int,
        plainText: String,
        paragraph: ParagraphBoxProvider,
    ): CaretMetrics? {
        val plainTextLength = plainText.length
        if (plainTextLength == 0 || offset < 0 || offset > plainTextLength) {
            return null
        }

        val safeOffset = offset.coerceIn(0, plainTextLength)
        val prevCodeUnit = plainText[max(0, safeOffset - 1)].code
        val nextCodeUnit = plainText.getOrNull(safeOffset)?.code

        val needsSearch = needsGraphemeExtendedSearch(prevCodeUnit, nextCodeUnit)

        val boxes = findBoxesUpstream(
            paragraph = paragraph,
            offset = safeOffset,
            plainTextLength = plainTextLength,
            needsSearch = needsSearch,
            stopAtNewline = prevCodeUnit == NEWLINE_CODE_UNIT,
        )
        if (boxes.isEmpty()) return null

        val box = boxes.last()
        return if (prevCodeUnit == NEWLINE_CODE_UNIT) {
            EmptyLineCaretMetrics(lineHorizontalOffset = box.right)
        } else {
            LineCaretMetrics(
                offset = Offset(x = box.left, y = box.bottom),
                fullWidth = box.right - box.left,
            )
        }
    }

    private fun getMetricsFromDownstream(
        offset: Int,
        plainText: String,
        paragraph: ParagraphBoxProvider,
    ): CaretMetrics? {
        val plainTextLength = plainText.length
        if (plainTextLength == 0 || offset < 0 || offset > plainTextLength) {
            return null
        }

        val safeOffset = offset.coerceIn(0, plainTextLength)
        val nextCodeUnit = plainText[min(safeOffset, plainTextLength - 1)].code

        val needsSearch = needsGraphemeExtendedSearch(nextCodeUnit, null)
        val boxes = findBoxesDownstream(
            paragraph = paragraph,
            offset = safeOffset,
            plainTextLength = plainTextLength,
            needsSearch = needsSearch,
        )

        if (boxes.isEmpty()) return null

        val box = boxes.first()
        return LineCaretMetrics(
            offset = Offset(x = box.left, y = box.top),
            fullWidth = box.right - box.left,
        )
    }

    private fun findBoxesUpstream(
        paragraph: ParagraphBoxProvider,
        offset: Int,
        plainTextLength: Int,
        needsSearch: Boolean,
        stopAtNewline: Boolean,
    ): List<Rect> {
        var graphemeLength = if (needsSearch) 2 else 1

        while (true) {
            val rangeStart = max(0, offset - graphemeLength)
            val boxes = paragraph.getBoxesForRange(rangeStart, offset)
            if (boxes.isNotEmpty()) {
                return boxes
            }

            if (!needsSearch || stopAtNewline || rangeStart == 0) {
                return emptyList()
            }

            val nextLength = min(plainTextLength, graphemeLength * 2)
            if (nextLength == graphemeLength) {
                return emptyList()
            }
            graphemeLength = nextLength
        }
    }

    private fun findBoxesDownstream(
        paragraph: ParagraphBoxProvider,
        offset: Int,
        plainTextLength: Int,
        needsSearch: Boolean,
    ): List<Rect> {
        var graphemeLength = if (needsSearch) 2 else 1
        val maxSearchEnd = plainTextLength shl 1

        while (true) {
            val rangeEnd = offset + graphemeLength
            val boxes = paragraph.getBoxesForRange(offset, rangeEnd)
            if (boxes.isNotEmpty()) {
                return boxes
            }

            if (!needsSearch || rangeEnd >= maxSearchEnd) {
                return emptyList()
            }

            val nextLength = graphemeLength * 2
            if (nextLength == graphemeLength) {
                return emptyList()
            }
            graphemeLength = nextLength
        }
    }

    private fun needsGraphemeExtendedSearch(codeUnit: Int, codeUnitAfter: Int?): Boolean {
        return MongolTextTools.isHighSurrogate(codeUnit) || MongolTextTools.isLowSurrogate(codeUnit) || codeUnitAfter == ZERO_WIDTH_JOINER_CODE_UNIT || MongolTextTools.isUnicodeDirectionality(
            codeUnit
        )
    }
}
